Userland/Libraries/LibPDF/Renderer.cpp

```diff
@@ -1205,6 +1206,49 @@ PDFErrorOr<Renderer::LoadedImage> Renderer::load_image(NonnullRefPtr<StreamObjec
     return LoadedImage { bitmap, is_image_mask };
 }
 
+PDFErrorOr<NonnullRefPtr<Gfx::Bitmap>> Renderer::make_mask_bitmap_from_array(NonnullRefPtr<ArrayObject> array, NonnullRefPtr<StreamObject> image)
+{
+    // PDF 1.7 spec, 4.8.5. Masked Images, Color Key Masking
+    // "For color key masking, the value of the Mask entry is an array of 2 × n integers, [min_1 max_1 ... min_n max_n],
+    //  where n is the number of color components in the image’s color space. Each integer must be in the range 0 to 2**(BitsPerComponent − 1),
+    //  representing color values _before_ decoding with the Decode array.
+    //  An image sample is masked [...] if min_i ≤ c_i ≤ max_i for all 1 ≤ i ≤ n."
+    // For indexed images, this means the array masks the index, not the color.
+    auto image_dict = image->dict();
+    auto width = TRY(m_document->resolve_to<int>(image_dict->get_value(CommonNames::Width)));
+    auto height = TRY(m_document->resolve_to<int>(image_dict->get_value(CommonNames::Height)));
+    auto bits_per_component = TRY(m_document->resolve_to<int>(image_dict->get_value(CommonNames::BitsPerComponent)));
+    VERIFY(bits_per_component == 1 || bits_per_component == 2 || bits_per_component == 4 || bits_per_component == 8 || bits_per_component == 16);
+
+    if (array->size() % 2 != 0)
+        return Error(Error::Type::MalformedPDF, "Mask array must have an even number of elements");
+    auto n_components = array->size() / 2;
+    Vector<int, 4> min, max;
+    for (size_t i = 0; i < n_components; ++i) {
+        min.append(array->at(i * 2).to_int());
+        max.append(array->at(i * 2 + 1).to_int());
+    }
+
+    auto mask_bitmap = MUST(Gfx::Bitmap::create(Gfx::BitmapFormat::BGRA8888, { width, height }));
+    auto bit_stream = make<BigEndianInputBitStream>(make<FixedMemoryStream>(image->bytes()));
+    for (int y = 0; y < height; ++y) {
+        for (int x = 0; x < width; ++x) {
+            bool is_masked = true;
+            for (size_t i = 0; i < n_components; ++i) {
+                u16 sample = TRY(bit_stream->read_bits(bits_per_component));
+                if (sample < min[i] || sample > max[i]) {
+                    is_masked = false;
+                    TRY(bit_stream->read_bits((n_components - 1 - i) * bits_per_component));
+                    break;
+                }
+            }
+            mask_bitmap->set_pixel(x, y, Color::from_argb(is_masked ? 0x00000000 : 0xffffffff));
```

Comment:
We already have `NamedColor::Transparent` for `0x00000000`, maybe we can introduce `NamedColor::Opaque` for `0xxxxxxxff`?
Or if you don't want to add these to `NamedColor`, it would be nice to at least have local variable for these magic values.

Reply:
They're not magic, they're all ones and all zeroes. That's like `int const one = 1;` 😢

I added channel separators, does that help?

Reply:
Not really, but do as you want